include/tvm/relax/nested_msg.h

```diff
@@ -305,6 +305,101 @@ NestedMsg<T> MapToNestedMsg(StructInfo sinfo, FType fmapleaf) {
   }
 }
 
+/*!
+ * \brief Map expr with possible nested-tuple to nested message,
+ * according to its struct info.
+ *
+ * This function will unpack recursive sinfo, run fstepproces in every
+ * recursion step, and run fmapleaf for each leaf.
+ * then recursively combines the results together into a NestedMsg.
+ *
+ * The nesting structure will corresponds to the tuple structure.
+ *
+ * \param expr The input expression with tuple struct info.
+ * \param fstepprocess The processing function for each step with signature
+ *             Expr fstepprocess(Expr, size_t)
+ * \param fmapleaf The mapping function for each leaf with signature
+ *             NestedMsg<T> fmap(Expr)
+ * \tparam T the content type of nested msg
+ * \tparam FType The mapping function type
+ */
+template <typename T, typename FType1, typename FType2>
+NestedMsg<T> MapToNestedMsg(Expr expr, FType1 fstepprocess, FType2 fmapleaf) {
+  auto sinfo = GetStructInfo(expr);
+  if (auto* tuple = sinfo.as<TupleStructInfoNode>()) {
+    Array<NestedMsg<T>> res;
+    res.reserve(tuple->fields.size());
+    for (size_t i = 0; i < tuple->fields.size(); ++i) {
+      res.push_back(
+          MapToNestedMsg<T, FType1, FType2>(fstepprocess(expr, i), fstepprocess, fmapleaf));
+    }
+    return res;
+  } else {
+    return fmapleaf(expr);
+  }
+}
+
+/*!
+ * \brief Map expr with possible nested-tuple to nested message,
+ * according to its struct info.
+ */
+template <typename T, typename FType>
+NestedMsg<T> MapToNestedMsgNew(Expr expr, FType fmapleaf) {
```

Comment:
Where is the use of this func?

Reply:
It is just MapToNestedMsg when `fstepprocess` is set to `GetField` (choose `expr->fields[i]` or `TupleGetItem(expr, i)` on occation), just like TransformTupleLeaf :)

Reply:
But what should be the name of the new MapToNestedMsg?

We already have a `MapToNestedMsg(Expr expr, FType fmapleaf)` that recursives according to `expr->as<TupleNode>`. The new MapToNestedMsg seems to have the same signature as the old one.

Reply:
Removed and merged into MapToNestedMsg. Plz see the current impl.